lib/Sema/CSSimplify.cpp

```diff
@@ -3408,6 +3435,8 @@ ConstraintSystem::matchFunctionTypes(FunctionType *func1, FunctionType *func2,
       if (!anchor)
         return getTypeMatchFailure(argumentLocator);
 
+      loc = adjustFunctionCoercionFixLocator(*this, loc, loc->getPath());
```

Comment:
I'm afraid that we'd end up with this adjustment all over the place. Have you tried a slightly different approach where `CoerceExpr` stays as an anchor but instead of `FunctionCoercion` we add a `CoercionOperand` element and that forces `simplifyLocator` to always produce underlying expression? This way it should be possible to add that element to conversion locator during CSGen and that would do the trick since ambiguity and other diagnostics try to simply?

Reply:
Yeah, that is kind of what I had in mind initially but haven't tried because somehow interpreted from the last comment we want to adjust only before recording the fix, but let me give this a try. And sorry for all the back and forth

Reply:
No need to apologize! I though that we could do the switch to the underlying element in one place i.e. in matchTypes somewhere but having looked at your changes it seems like it might be easier to just do it during constraint generation...

Reply:
Just an update, working on this is just slowly... a bit swamped with work stuff this past/next few weeks. Hopefully will make some progress soon :)

Reply:
Sounds great, thank you!

Reply:
Just updated, let me know what you think and if I missed something along the way. This didn't make things as simple as I imagined it would be, but felt a little bit better than previous "fix adjustment approach".